SonarQube.TeamBuild.PreProcessor/Roslyn/AnalyzerInstaller.cs

```diff
@@ -18,7 +62,142 @@ public class AnalyzerInstaller : IAnalyzerInstaller
+            this.remoteRepository = PackageRepositoryFactory.Default.CreateRepository(this.packageSource);
+
+            // Create the local NuGet cache
+            Directory.CreateDirectory(this.localCacheDirectory);
+            this.manager = new PackageManager(remoteRepository, this.localCacheDirectory);
+            this.manager.Logger = new NuGetLoggerAdapter(this.logger);
+            
+            ISet<string> assemblyPaths = new HashSet<string>();
+
+            foreach(NuGetPackageInfo package in packages)
+            {
+                ProcessPackage(package, assemblyPaths);
+            }
+
+            this.logger.LogInfo(RoslynResources.PackagesInstalled);
+
+            return assemblyPaths;
+        }
+
+        #endregion
+
+        #region Private methods
+
+        /// <summary>
+        /// We want the NuGet cache to be in a well-known location so we can re-use packages that have
+        /// already been installed (although this won't help for e.g. hosted build agents
+        /// </summary>
+        private static string GetLocalCacheDirectory()
+        {
+            string localCache = Path.Combine(Path.GetTempPath(), ".sonarqube", ".nuget");
+            return localCache;
+        }
+
+        private void ProcessPackage(NuGetPackageInfo packageInfo, ISet<string> assemblyPaths)
+        {
+            Debug.Assert(packageInfo != null, "Not expecting the package info to be null");
+
+            IPackage package = FetchPackage(packageInfo.Id, packageInfo.Version);
+            if (package != null)
+            {
+                IEnumerable<string> filesFromPackage = GetFilesFromPackage(package);
+                assemblyPaths.AddRange(filesFromPackage);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to download a NuGet package with the specified id and version
+        /// </summary>
+        private IPackage FetchPackage(string packageId, string packageVersion)
+        {
+            SemanticVersion version = new SemanticVersion(packageVersion);
+
+            try
+            {
+                logger.LogDebug(RoslynResources.LocatingPackage, packageId, version);
+                
+                // "InstallPackage" is a no-op if the package already exists in the local cache
+                this.manager.InstallPackage(packageId, version, ignoreDependencies: false, allowPrereleaseVersions: true);
+                this.logger.LogDebug(RoslynResources.PackageInstalled);
+            }
+            catch (InvalidOperationException e)
+            {
+                logger.LogError(RoslynResources.PackageInstallFailed, e.Message);
+                throw;
+            }
+
+            IPackage package = this.manager.LocalRepository.FindPackage(packageId, version);
+            Debug.Assert(package != null, "Package should be located as it has just been installed");
+
+            return package;
+        }
+        
+        private IEnumerable<string> GetFilesFromPackage(IPackage package)
```

Comment:
This is simpler and cleaner than the approach used in the SDK code. We could fix up the SDK code, but it's not urgent because it works.